jmclient/jmclient/wallet.py

```diff
@@ -1593,6 +1610,172 @@ def _get_mixdepth_from_path(self, path):
+            raise ValueError()
+        day_and_shorter_tuple = (dt.day, dt.hour, dt.minute, dt.second, dt.microsecond)
+        if day_and_shorter_tuple != cls.TIMELOCK_DAY_AND_SHORTER:
+            raise ValueError()
+        timenumber = (dt.year - cls.TIMELOCK_EPOCH_YEAR)*(cls.MONTHS_IN_YEAR //
+            cls.TIMENUMBER_UNIT) + ((dt.month - cls.TIMELOCK_EPOCH_MONTH) // cls.TIMENUMBER_UNIT)
+        if timenumber < 0 or timenumber > cls.TIMENUMBERS_PER_PUBKEY:
+            raise ValueError("datetime out of range")
+        return timenumber
+
+    @classmethod
+    def _is_timelocked_path(cls, path):
+        #return path[2] == cls.BIP32_TIMELOCK_ID and isinstance(path[-1], datetime)
+        return isinstance(path[-1], datetime)
+
+    @classmethod
+    def _is_burn_path(cls, path):
+        #TODO this doesnt work, because its hardened so the number is + 2*32 i think
+        return path[2] == cls.BIP32_BURN_ID
+
+    def _populate_script_map(self):
+        super(FidelityBondMixin, self)._populate_script_map()
+        for md in self._index_cache:
+            address_type = self.BIP32_TIMELOCK_ID
+            ##BURN_ID not done yet
+            for i in range(self._index_cache[md][address_type]):
+                for timenumber in range(TIMENUMBERS_PER_PUBKEY):
+                    path = self.get_path(md, address_type, i, timenumber)
+                    script = self.get_script_from_path(path)
+                    self._script_map[script] = path
+
+    @classmethod
+    def _get_supported_address_types(cls):
+        return (cls.BIP32_EXT_ID, cls.BIP32_INT_ID, cls.BIP32_TIMELOCK_ID, cls.BIP32_BURN_ID)
+
+    def _get_priv_from_path(self, path):
+        if self._is_timelocked_path(path):
+            key_path = path[:-1]
+            locktime = timegm(path[-1].timetuple())
+            engine = ENGINES[TYPE_TIMELOCK_P2WSH]
+            privkey = engine.derive_bip32_privkey(self._master_key, key_path)
+            return (privkey, locktime), engine
+        elif self._is_burn_path(path):
+            raise Exception("not implemented yet")
+        else:
+            return super(FidelityBondMixin, self)._get_priv_from_path(path)
+
+    def get_path(self, mixdepth=None, address_type=None, index=None, timenumber=None):
+        if address_type == None or address_type in (self.BIP32_EXT_ID, self.BIP32_INT_ID,
+                self.BIP32_BURN_ID) or index == None:
+            return super(FidelityBondMixin, self).get_path(mixdepth, address_type, index)
+        elif address_type == self.BIP32_TIMELOCK_ID:
+            assert timenumber != None
+            timestamp = self._time_number_to_timestamp(timenumber)
+            return tuple(chain(self._get_bip32_export_path(mixdepth, address_type),
+                               (index, timestamp)))
+        else:
+            assert 0
+
+    """
+    We define a new serialization of the bip32 path to include bip65 timelock addresses
+    Previously it was m/44'/0'/3/0
```

Comment:
Is this missing a level? it should be purpose/coin type/account/address-type(your parlance)/index, so when purpose is per BIP49 you get m/49'/0'/0'/address type/index .. '3' here is the address type, right?

Reply:
Yes you're right, will fix